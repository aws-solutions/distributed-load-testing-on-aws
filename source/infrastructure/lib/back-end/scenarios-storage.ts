// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: Apache-2.0

import { RemovalPolicy } from "aws-cdk-lib";
import { AttributeType, BillingMode, Table, TableEncryption } from "aws-cdk-lib/aws-dynamodb";
import { AnyPrincipal, Effect, Policy, PolicyStatement } from "aws-cdk-lib/aws-iam";
import { BlockPublicAccess, Bucket, BucketEncryption, HttpMethods } from "aws-cdk-lib/aws-s3";
import { Construct } from "constructs";

export interface ScenarioTestRunnerStorageConstructProps {
  // S3 Logs Bucket
  readonly s3LogsBucket: Bucket;
  // CloudFront domain name
  readonly cloudFrontDomainName: string;
  // Solution Id
  readonly solutionId: string;
}

/**
 * Distributed Load Testing storage construct
 * Creates an S3 bucket to store test scenarios and
 * a Dynamodb table to store tests and test configuration
 */
export class ScenarioTestRunnerStorageConstruct extends Construct {
  public scenariosBucket: Bucket;
  public scenariosS3Policy: Policy;
  public scenariosTable: Table;
  public scenarioDynamoDbPolicy: Policy;
  public historyTable: Table;
  public historyDynamoDbPolicy: Policy;

  constructor(scope: Construct, id: string, props: ScenarioTestRunnerStorageConstructProps) {
    super(scope, id);

    this.scenariosBucket = new Bucket(this, "DLTScenariosBucket", {
      removalPolicy: RemovalPolicy.RETAIN,
      serverAccessLogsBucket: props.s3LogsBucket,
      serverAccessLogsPrefix: "scenarios-bucket-access/",
      encryption: BucketEncryption.KMS_MANAGED,
      enforceSSL: true,
      versioned: true,
      blockPublicAccess: BlockPublicAccess.BLOCK_ALL,
      cors: [
        {
          allowedMethods: [HttpMethods.GET, HttpMethods.POST, HttpMethods.PUT],
          allowedOrigins: [`https://${props.cloudFrontDomainName}`],
          allowedHeaders: ["*"],
          exposedHeaders: ["ETag"],
        },
      ],
    });

    this.scenariosBucket.addToResourcePolicy(
      new PolicyStatement({
        actions: ["s3:*"],
        resources: [this.scenariosBucket.bucketArn, `${this.scenariosBucket.bucketArn}/*`],
        effect: Effect.DENY,
        principals: [new AnyPrincipal()],
        conditions: {
          Bool: {
            "aws:SecureTransport": false,
          },
        },
      })
    );

    this.scenariosS3Policy = new Policy(this, "ScenariosS3Policy", {
      statements: [
        new PolicyStatement({
          effect: Effect.ALLOW,
          actions: ["s3:HeadObject", "s3:PutObject", "s3:GetObject", "s3:DeleteObject", "s3:ListBucket"],
          resources: [this.scenariosBucket.bucketArn, `${this.scenariosBucket.bucketArn}/*`],
        }),
      ],
    });

    this.scenariosTable = new Table(this, "DLTScenariosTable", {
      billingMode: BillingMode.PAY_PER_REQUEST,
      encryption: TableEncryption.AWS_MANAGED,
      partitionKey: { name: "testId", type: AttributeType.STRING },
      pointInTimeRecovery: true,
    });

    this.historyTable = new Table(this, "DLTHistoryTable", {
      billingMode: BillingMode.PAY_PER_REQUEST,
      encryption: TableEncryption.AWS_MANAGED,
      partitionKey: { name: "testId", type: AttributeType.STRING },
      sortKey: { name: "testRunId", type: AttributeType.STRING },
      pointInTimeRecovery: true,
    });

    const historyDDBActions = ["dynamodb:BatchWriteItem", "dynamodb:PutItem", "dynamodb:Query"];
    this.historyDynamoDbPolicy = new Policy(this, "HistoryDynamoDbPolicy", {
      statements: [
        new PolicyStatement({
          effect: Effect.ALLOW,
          actions: historyDDBActions,
          resources: [this.historyTable.tableArn],
        }),
      ],
    });

    const scenariosDDBActions = [
      "dynamodb:DeleteItem",
      "dynamodb:GetItem",
      "dynamodb:PutItem",
      "dynamodb:Scan",
      "dynamodb:UpdateItem",
    ];
    this.scenarioDynamoDbPolicy = new Policy(this, "ScenarioDynamoDbPolicy", {
      statements: [
        new PolicyStatement({
          effect: Effect.ALLOW,
          actions: scenariosDDBActions,
          resources: [this.scenariosTable.tableArn],
        }),
      ],
    });
  }
}
