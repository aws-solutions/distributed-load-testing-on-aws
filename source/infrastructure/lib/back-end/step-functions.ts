// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: Apache-2.0

import {
  Chain,
  Choice,
  Condition,
  Fail,
  Pass,
  LogLevel,
  Map as SFMap,
  StateMachine,
  Succeed,
  Wait,
  WaitTime,
  JsonPath,
  DefinitionBody,
  TaskInput,
} from "aws-cdk-lib/aws-stepfunctions";
import { LambdaInvoke } from "aws-cdk-lib/aws-stepfunctions-tasks";
import { Aws, CfnResource, Duration } from "aws-cdk-lib";
import { Policy } from "aws-cdk-lib/aws-iam";
import { LogGroup, RetentionDays } from "aws-cdk-lib/aws-logs";
import { IFunction } from "aws-cdk-lib/aws-lambda";
import { Construct } from "constructs";

export interface TaskRunnerStepFunctionConstructProps {
  // State machine Lambda functions
  readonly taskStatusChecker: IFunction;
  readonly taskRunner: IFunction;
  readonly resultsParser: IFunction;
  readonly taskCanceler: IFunction;
  // Solution ID
  readonly solutionId: string;
  readonly suffix: string;
}

/**
 * Creates the Step function state machine to control the Fargate tasks
 */
export class TaskRunnerStepFunctionConstruct extends Construct {
  public taskRunnerStepFunctions: StateMachine;

  constructor(scope: Construct, id: string, props: TaskRunnerStepFunctionConstructProps) {
    super(scope, id);

    const stepFunctionsLogGroup = new LogGroup(this, "StepFunctionsLogGroup", {
      retention: RetentionDays.ONE_YEAR,
      logGroupName: `/aws/vendedlogs/states/StepFunctionsLogGroup${Aws.STACK_NAME}${props.suffix}`,
    });
    const stepFunctionsLogGroupResource = stepFunctionsLogGroup.node.defaultChild as CfnResource;
    stepFunctionsLogGroupResource.addMetadata("cfn_nag", {
      rules_to_suppress: [
        {
          id: "W84",
          reason: "KMS encryption unnecessary for log group",
        },
      ],
    });

    const done = new Succeed(this, "Done");
    const mapEnd = new Pass(this, "Map End");
    const parseResult = new LambdaInvoke(this, "Parse result", {
      lambdaFunction: props.resultsParser,
      payload: TaskInput.fromObject({
        "testTaskConfig.$": "$.testTaskConfig",
        "testId.$": "$.testId",
        "testType.$": "$.testType",
        "fileType.$": "$.fileType",
        "showLive.$": "$.showLive",
        "testDuration.$": "$.testDuration",
        "prefix.$": "$.prefix",
        "executionStart.$": "$$.Execution.StartTime",
      }),
    });
    parseResult.next(done);

    const checkWorkerStatus = new LambdaInvoke(this, "Check worker status", {
      lambdaFunction: props.taskStatusChecker,
      inputPath: "$",
      outputPath: "$.Payload",
    });

    const checkTaskStatus = new LambdaInvoke(this, "Check task status", {
      lambdaFunction: props.taskStatusChecker,
      inputPath: "$",
      outputPath: "$.Payload",
    });

    const waitTask = new Wait(this, "Wait 1 minute - task status", {
      comment: "Wait 1 minute to check task status again",
      time: WaitTime.duration(Duration.seconds(60)),
    });
    waitTask.next(checkTaskStatus);

    const waitTestDuration = new Wait(this, "Wait specified test duration", {
      time: WaitTime.secondsPath("$.testDuration"),
    });
    waitTestDuration.next(checkTaskStatus);

    const allTasksDone = new Choice(this, "Are all tasks done?");
    allTasksDone.when(Condition.booleanEquals("$.isRunning", false), mapEnd);
    allTasksDone.otherwise(waitTask);

    checkTaskStatus.next(allTasksDone);

    const cancelTest = new LambdaInvoke(this, "Cancel Test", {
      lambdaFunction: props.taskCanceler,
      inputPath: "$",
      resultPath: JsonPath.DISCARD,
    });
    cancelTest.next(mapEnd);

    const waitWorker = new Wait(this, "Wait 1 minute - worker status", {
      comment: "Wait 1 minute to check task status again",
      time: WaitTime.duration(Duration.seconds(60)),
    });
    waitWorker.next(checkWorkerStatus);

    const regionConfigsForTest = new SFMap(this, "Regions for testing", {
      inputPath: "$",
      resultPath: JsonPath.DISCARD,
      itemsPath: "$.testTaskConfig",
      itemSelector: {
        "testTaskConfig.$": "$$.Map.Item.Value",
        "testId.$": "$.testId",
        "testType.$": "$.testType",
        "fileType.$": "$.fileType",
        "showLive.$": "$.showLive",
        "testDuration.$": "$.testDuration",
        "prefix.$": "$.prefix",
      },
    });

    const runWorkers = new LambdaInvoke(this, "Run workers", {
      lambdaFunction: props.taskRunner,
      inputPath: "$",
      outputPath: "$.Payload",
    });

    const requiresLeader = new Choice(this, "Requires leader?");
    requiresLeader.when(Condition.booleanEquals("$.isRunning", false), cancelTest);
    requiresLeader.when(Condition.isNotPresent("$.taskIds"), waitTestDuration);
    requiresLeader.otherwise(waitWorker);

    runWorkers.next(requiresLeader);

    const runLeaderTask = new LambdaInvoke(this, "Run leader task", {
      lambdaFunction: props.taskRunner,
      inputPath: "$",
      outputPath: "$.Payload",
    });
    runLeaderTask.addCatch(cancelTest, { resultPath: "$.error" });

    runLeaderTask.next(waitTestDuration);

    const allWorkersRunning = new Choice(this, "Are all workers running?");
    allWorkersRunning.when(Condition.booleanEquals("$.isRunning", false), cancelTest);
    allWorkersRunning.when(Condition.numberEqualsJsonPath("$.numTasksRunning", "$.numTasksTotal"), runLeaderTask);
    allWorkersRunning.otherwise(waitWorker);

    checkWorkerStatus.next(allWorkersRunning);

    const testIsStillRunning = new Fail(this, "Test is still running", {
      cause: "The same test is already running.",
      error: "TestAlreadyRunning",
    });

    const noRunningTests = new Choice(this, "No running tests");
    noRunningTests.when(Condition.booleanEquals("$.isRunning", false), runWorkers);
    noRunningTests.otherwise(testIsStillRunning);

    const checkRunningTests = new LambdaInvoke(this, "Check running tests", {
      lambdaFunction: props.taskStatusChecker,
      inputPath: "$",
      outputPath: "$.Payload",
    });
    checkRunningTests.next(noRunningTests);

    const definition = Chain.start(regionConfigsForTest.itemProcessor(checkRunningTests)).next(parseResult);

    this.taskRunnerStepFunctions = new StateMachine(this, "TaskRunnerStepFunctions", {
      definitionBody: DefinitionBody.fromChainable(definition),
      logs: {
        destination: stepFunctionsLogGroup,
        level: LogLevel.ALL,
        includeExecutionData: false,
      },
    });
    const stepFunctionsRoleResource = this.taskRunnerStepFunctions.role.node.defaultChild as CfnResource;
    stepFunctionsRoleResource.addMetadata("cfn_nag", {
      rules_to_suppress: [
        {
          id: "W11",
          reason: "CloudWatch logs actions do not support resource level permissions",
        },
        {
          id: "W12",
          reason: "CloudWatch logs actions do not support resource level permissions",
        },
      ],
    });
    const stepFunctionPolicy = this.taskRunnerStepFunctions.role.node.findChild("DefaultPolicy") as Policy;
    const policyResource = stepFunctionPolicy.node.defaultChild as CfnResource;
    policyResource.addMetadata("cfn_nag", {
      rules_to_suppress: [
        {
          id: "W12",
          reason: "CloudWatch logs actions do not support resource level permissions",
        },
        {
          id: "W76",
          reason: "The IAM policy is written for least-privilege access.",
        },
      ],
    });
  }
}
