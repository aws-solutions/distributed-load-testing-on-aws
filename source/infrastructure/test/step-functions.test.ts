// Copyright 2021 Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: Apache-2.0

import { Template } from "aws-cdk-lib/assertions";
import { App, DefaultStackSynthesizer, Stack } from "aws-cdk-lib";
import { TaskRunnerStepFunctionConstruct } from "../lib/back-end/step-functions";
import { Code, Function as LambdaFunction, Runtime } from "aws-cdk-lib/aws-lambda";
import { Bucket } from "aws-cdk-lib/aws-s3";
import { Effect, PolicyDocument, PolicyStatement, Role, ServicePrincipal } from "aws-cdk-lib/aws-iam";

test("DLT API Test", () => {
  const app = new App();
  const stack = new Stack(app, "DLTStack", {
    synthesizer: new DefaultStackSynthesizer({
      generateBootstrapVersionRule: false,
    }),
  });

  const testRole = new Role(stack, "TestRole", {
    assumedBy: new ServicePrincipal("lambda.amazonaws.com"),
    inlinePolicies: {
      DenyPolicy: new PolicyDocument({
        statements: [
          new PolicyStatement({
            effect: Effect.DENY,
            actions: ["*"],
            resources: ["*"],
          }),
        ],
      }),
    },
  });

  const testLambda = new LambdaFunction(stack, "TestFunction", {
    code: Code.fromBucket(Bucket.fromBucketName(stack, "SourceCodeBucket", "testbucket"), "custom-resource.zip"),
    handler: "index.handler",
    runtime: Runtime.NODEJS_18_X,
    role: testRole,
  });

  const testStateMachine = new TaskRunnerStepFunctionConstruct(stack, "TaskRunnerStepFunction", {
    taskStatusChecker: testLambda,
    taskRunner: testLambda,
    resultsParser: testLambda,
    taskCanceler: testLambda,
    solutionId: "testId",
    suffix: "abc-def-xyz",
  });

  expect(Template.fromStack(stack)).toMatchSnapshot();
  expect(testStateMachine.taskRunnerStepFunctions).toBeDefined();
});
