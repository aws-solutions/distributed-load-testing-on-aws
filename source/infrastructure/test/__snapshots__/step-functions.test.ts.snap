// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DLT API Test 1`] = `
{
  "Mappings": {
    "ServiceprincipalMap": {
      "af-south-1": {
        "states": "states.af-south-1.amazonaws.com",
      },
      "ap-east-1": {
        "states": "states.ap-east-1.amazonaws.com",
      },
      "ap-northeast-1": {
        "states": "states.ap-northeast-1.amazonaws.com",
      },
      "ap-northeast-2": {
        "states": "states.ap-northeast-2.amazonaws.com",
      },
      "ap-northeast-3": {
        "states": "states.ap-northeast-3.amazonaws.com",
      },
      "ap-south-1": {
        "states": "states.ap-south-1.amazonaws.com",
      },
      "ap-south-2": {
        "states": "states.ap-south-2.amazonaws.com",
      },
      "ap-southeast-1": {
        "states": "states.ap-southeast-1.amazonaws.com",
      },
      "ap-southeast-2": {
        "states": "states.ap-southeast-2.amazonaws.com",
      },
      "ap-southeast-3": {
        "states": "states.ap-southeast-3.amazonaws.com",
      },
      "ap-southeast-4": {
        "states": "states.ap-southeast-4.amazonaws.com",
      },
      "ca-central-1": {
        "states": "states.ca-central-1.amazonaws.com",
      },
      "cn-north-1": {
        "states": "states.cn-north-1.amazonaws.com",
      },
      "cn-northwest-1": {
        "states": "states.cn-northwest-1.amazonaws.com",
      },
      "eu-central-1": {
        "states": "states.eu-central-1.amazonaws.com",
      },
      "eu-central-2": {
        "states": "states.eu-central-2.amazonaws.com",
      },
      "eu-north-1": {
        "states": "states.eu-north-1.amazonaws.com",
      },
      "eu-south-1": {
        "states": "states.eu-south-1.amazonaws.com",
      },
      "eu-south-2": {
        "states": "states.eu-south-2.amazonaws.com",
      },
      "eu-west-1": {
        "states": "states.eu-west-1.amazonaws.com",
      },
      "eu-west-2": {
        "states": "states.eu-west-2.amazonaws.com",
      },
      "eu-west-3": {
        "states": "states.eu-west-3.amazonaws.com",
      },
      "il-central-1": {
        "states": "states.il-central-1.amazonaws.com",
      },
      "me-central-1": {
        "states": "states.me-central-1.amazonaws.com",
      },
      "me-south-1": {
        "states": "states.me-south-1.amazonaws.com",
      },
      "sa-east-1": {
        "states": "states.sa-east-1.amazonaws.com",
      },
      "us-east-1": {
        "states": "states.us-east-1.amazonaws.com",
      },
      "us-east-2": {
        "states": "states.us-east-2.amazonaws.com",
      },
      "us-gov-east-1": {
        "states": "states.us-gov-east-1.amazonaws.com",
      },
      "us-gov-west-1": {
        "states": "states.us-gov-west-1.amazonaws.com",
      },
      "us-iso-east-1": {
        "states": "states.amazonaws.com",
      },
      "us-iso-west-1": {
        "states": "states.amazonaws.com",
      },
      "us-isob-east-1": {
        "states": "states.amazonaws.com",
      },
      "us-west-1": {
        "states": "states.us-west-1.amazonaws.com",
      },
      "us-west-2": {
        "states": "states.us-west-2.amazonaws.com",
      },
    },
  },
  "Resources": {
    "TaskRunnerStepFunctionStepFunctionsLogGroup39F53B04": {
      "DeletionPolicy": "Retain",
      "Metadata": {
        "cfn_nag": {
          "rules_to_suppress": [
            {
              "id": "W84",
              "reason": "KMS encryption unnecessary for log group",
            },
          ],
        },
      },
      "Properties": {
        "LogGroupName": {
          "Fn::Join": [
            "",
            [
              "/aws/vendedlogs/states/StepFunctionsLogGroup",
              {
                "Ref": "AWS::StackName",
              },
              "abc-def-xyz",
            ],
          ],
        },
        "RetentionInDays": 365,
      },
      "Type": "AWS::Logs::LogGroup",
      "UpdateReplacePolicy": "Retain",
    },
    "TaskRunnerStepFunctionTaskRunnerStepFunctionsC50015AC": {
      "DeletionPolicy": "Delete",
      "DependsOn": [
        "TaskRunnerStepFunctionTaskRunnerStepFunctionsRoleDefaultPolicyACCEC854",
        "TaskRunnerStepFunctionTaskRunnerStepFunctionsRole706897AC",
      ],
      "Properties": {
        "DefinitionString": {
          "Fn::Join": [
            "",
            [
              "{"StartAt":"Regions for testing","States":{"Regions for testing":{"Type":"Map","ResultPath":null,"Next":"Parse result","InputPath":"$","Parameters":{"testTaskConfig.$":"$$.Map.Item.Value","testId.$":"$.testId","testType.$":"$.testType","fileType.$":"$.fileType","showLive.$":"$.showLive","testDuration.$":"$.testDuration","prefix.$":"$.prefix"},"Iterator":{"StartAt":"Check running tests","States":{"Check running tests":{"Next":"No running tests","Retry":[{"ErrorEquals":["Lambda.ClientExecutionTimeoutException","Lambda.ServiceException","Lambda.AWSLambdaException","Lambda.SdkClientException"],"IntervalSeconds":2,"MaxAttempts":6,"BackoffRate":2}],"Type":"Task","InputPath":"$","OutputPath":"$.Payload","Resource":"arn:",
              {
                "Ref": "AWS::Partition",
              },
              ":states:::lambda:invoke","Parameters":{"FunctionName":"",
              {
                "Fn::GetAtt": [
                  "TestFunction22AD90FC",
                  "Arn",
                ],
              },
              "","Payload.$":"$"}},"No running tests":{"Type":"Choice","Choices":[{"Variable":"$.isRunning","BooleanEquals":false,"Next":"Run workers"}],"Default":"Test is still running"},"Test is still running":{"Type":"Fail","Error":"TestAlreadyRunning","Cause":"The same test is already running."},"Run workers":{"Next":"Requires leader?","Retry":[{"ErrorEquals":["Lambda.ClientExecutionTimeoutException","Lambda.ServiceException","Lambda.AWSLambdaException","Lambda.SdkClientException"],"IntervalSeconds":2,"MaxAttempts":6,"BackoffRate":2}],"Type":"Task","InputPath":"$","OutputPath":"$.Payload","Resource":"arn:",
              {
                "Ref": "AWS::Partition",
              },
              ":states:::lambda:invoke","Parameters":{"FunctionName":"",
              {
                "Fn::GetAtt": [
                  "TestFunction22AD90FC",
                  "Arn",
                ],
              },
              "","Payload.$":"$"}},"Requires leader?":{"Type":"Choice","Choices":[{"Variable":"$.isRunning","BooleanEquals":false,"Next":"Cancel Test"},{"Variable":"$.taskIds","IsPresent":false,"Next":"Wait specified test duration"}],"Default":"Wait 1 minute - worker status"},"Wait 1 minute - worker status":{"Type":"Wait","Comment":"Wait 1 minute to check task status again","Seconds":60,"Next":"Check worker status"},"Check worker status":{"Next":"Are all workers running?","Retry":[{"ErrorEquals":["Lambda.ClientExecutionTimeoutException","Lambda.ServiceException","Lambda.AWSLambdaException","Lambda.SdkClientException"],"IntervalSeconds":2,"MaxAttempts":6,"BackoffRate":2}],"Type":"Task","InputPath":"$","OutputPath":"$.Payload","Resource":"arn:",
              {
                "Ref": "AWS::Partition",
              },
              ":states:::lambda:invoke","Parameters":{"FunctionName":"",
              {
                "Fn::GetAtt": [
                  "TestFunction22AD90FC",
                  "Arn",
                ],
              },
              "","Payload.$":"$"}},"Are all workers running?":{"Type":"Choice","Choices":[{"Variable":"$.isRunning","BooleanEquals":false,"Next":"Cancel Test"},{"Variable":"$.numTasksRunning","NumericEqualsPath":"$.numTasksTotal","Next":"Run leader task"}],"Default":"Wait 1 minute - worker status"},"Cancel Test":{"Next":"Map End","Retry":[{"ErrorEquals":["Lambda.ClientExecutionTimeoutException","Lambda.ServiceException","Lambda.AWSLambdaException","Lambda.SdkClientException"],"IntervalSeconds":2,"MaxAttempts":6,"BackoffRate":2}],"Type":"Task","InputPath":"$","ResultPath":null,"Resource":"arn:",
              {
                "Ref": "AWS::Partition",
              },
              ":states:::lambda:invoke","Parameters":{"FunctionName":"",
              {
                "Fn::GetAtt": [
                  "TestFunction22AD90FC",
                  "Arn",
                ],
              },
              "","Payload.$":"$"}},"Run leader task":{"Next":"Wait specified test duration","Retry":[{"ErrorEquals":["Lambda.ClientExecutionTimeoutException","Lambda.ServiceException","Lambda.AWSLambdaException","Lambda.SdkClientException"],"IntervalSeconds":2,"MaxAttempts":6,"BackoffRate":2}],"Catch":[{"ErrorEquals":["States.ALL"],"ResultPath":"$.error","Next":"Cancel Test"}],"Type":"Task","InputPath":"$","OutputPath":"$.Payload","Resource":"arn:",
              {
                "Ref": "AWS::Partition",
              },
              ":states:::lambda:invoke","Parameters":{"FunctionName":"",
              {
                "Fn::GetAtt": [
                  "TestFunction22AD90FC",
                  "Arn",
                ],
              },
              "","Payload.$":"$"}},"Wait specified test duration":{"Type":"Wait","SecondsPath":"$.testDuration","Next":"Check task status"},"Check task status":{"Next":"Are all tasks done?","Retry":[{"ErrorEquals":["Lambda.ClientExecutionTimeoutException","Lambda.ServiceException","Lambda.AWSLambdaException","Lambda.SdkClientException"],"IntervalSeconds":2,"MaxAttempts":6,"BackoffRate":2}],"Type":"Task","InputPath":"$","OutputPath":"$.Payload","Resource":"arn:",
              {
                "Ref": "AWS::Partition",
              },
              ":states:::lambda:invoke","Parameters":{"FunctionName":"",
              {
                "Fn::GetAtt": [
                  "TestFunction22AD90FC",
                  "Arn",
                ],
              },
              "","Payload.$":"$"}},"Wait 1 minute - task status":{"Type":"Wait","Comment":"Wait 1 minute to check task status again","Seconds":60,"Next":"Check task status"},"Are all tasks done?":{"Type":"Choice","Choices":[{"Variable":"$.isRunning","BooleanEquals":false,"Next":"Map End"}],"Default":"Wait 1 minute - task status"},"Map End":{"Type":"Pass","End":true}}},"ItemsPath":"$.testTaskConfig"},"Parse result":{"Next":"Done","Retry":[{"ErrorEquals":["Lambda.ClientExecutionTimeoutException","Lambda.ServiceException","Lambda.AWSLambdaException","Lambda.SdkClientException"],"IntervalSeconds":2,"MaxAttempts":6,"BackoffRate":2}],"Type":"Task","Resource":"arn:",
              {
                "Ref": "AWS::Partition",
              },
              ":states:::lambda:invoke","Parameters":{"FunctionName":"",
              {
                "Fn::GetAtt": [
                  "TestFunction22AD90FC",
                  "Arn",
                ],
              },
              "","Payload.$":"$"}},"Done":{"Type":"Succeed"}}}",
            ],
          ],
        },
        "LoggingConfiguration": {
          "Destinations": [
            {
              "CloudWatchLogsLogGroup": {
                "LogGroupArn": {
                  "Fn::GetAtt": [
                    "TaskRunnerStepFunctionStepFunctionsLogGroup39F53B04",
                    "Arn",
                  ],
                },
              },
            },
          ],
          "IncludeExecutionData": false,
          "Level": "ALL",
        },
        "RoleArn": {
          "Fn::GetAtt": [
            "TaskRunnerStepFunctionTaskRunnerStepFunctionsRole706897AC",
            "Arn",
          ],
        },
      },
      "Type": "AWS::StepFunctions::StateMachine",
      "UpdateReplacePolicy": "Delete",
    },
    "TaskRunnerStepFunctionTaskRunnerStepFunctionsRole706897AC": {
      "Metadata": {
        "cfn_nag": {
          "rules_to_suppress": [
            {
              "id": "W11",
              "reason": "CloudWatch logs actions do not support resource level permissions",
            },
            {
              "id": "W12",
              "reason": "CloudWatch logs actions do not support resource level permissions",
            },
          ],
        },
      },
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": {
                  "Fn::FindInMap": [
                    "ServiceprincipalMap",
                    {
                      "Ref": "AWS::Region",
                    },
                    "states",
                  ],
                },
              },
            },
          ],
          "Version": "2012-10-17",
        },
      },
      "Type": "AWS::IAM::Role",
    },
    "TaskRunnerStepFunctionTaskRunnerStepFunctionsRoleDefaultPolicyACCEC854": {
      "Metadata": {
        "cfn_nag": {
          "rules_to_suppress": [
            {
              "id": "W12",
              "reason": "CloudWatch logs actions do not support resource level permissions",
            },
            {
              "id": "W76",
              "reason": "The IAM policy is written for least-privilege access.",
            },
          ],
        },
      },
      "Properties": {
        "PolicyDocument": {
          "Statement": [
            {
              "Action": [
                "logs:CreateLogDelivery",
                "logs:GetLogDelivery",
                "logs:UpdateLogDelivery",
                "logs:DeleteLogDelivery",
                "logs:ListLogDeliveries",
                "logs:PutResourcePolicy",
                "logs:DescribeResourcePolicies",
                "logs:DescribeLogGroups",
              ],
              "Effect": "Allow",
              "Resource": "*",
            },
            {
              "Action": "lambda:InvokeFunction",
              "Effect": "Allow",
              "Resource": [
                {
                  "Fn::GetAtt": [
                    "TestFunction22AD90FC",
                    "Arn",
                  ],
                },
                {
                  "Fn::Join": [
                    "",
                    [
                      {
                        "Fn::GetAtt": [
                          "TestFunction22AD90FC",
                          "Arn",
                        ],
                      },
                      ":*",
                    ],
                  ],
                },
              ],
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "TaskRunnerStepFunctionTaskRunnerStepFunctionsRoleDefaultPolicyACCEC854",
        "Roles": [
          {
            "Ref": "TaskRunnerStepFunctionTaskRunnerStepFunctionsRole706897AC",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "TestFunction22AD90FC": {
      "DependsOn": [
        "TestRole6C9272DF",
      ],
      "Properties": {
        "Code": {
          "S3Bucket": "testbucket",
          "S3Key": "custom-resource.zip",
        },
        "Handler": "index.handler",
        "Role": {
          "Fn::GetAtt": [
            "TestRole6C9272DF",
            "Arn",
          ],
        },
        "Runtime": "nodejs16.x",
      },
      "Type": "AWS::Lambda::Function",
    },
    "TestRole6C9272DF": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "Policies": [
          {
            "PolicyDocument": {
              "Statement": [
                {
                  "Action": "*",
                  "Effect": "Deny",
                  "Resource": "*",
                },
              ],
              "Version": "2012-10-17",
            },
            "PolicyName": "DenyPolicy",
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
  },
}
`;
